### PR TITLE
Wip/hughsie/genesys oss fuzz

### PR DESCRIPTION
**Pull request by hughsie**
Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [X] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

File: plugins/genesys/fu-genesys-usbhub-firmware.c [hunk 1]
```diff
@@ -336,50 +303,33 @@ fu_genesys_usbhub_firmware_parse(FuFirmware *firmware,
...
+							   error,
+							   FU_TYPE_GENESYS_USBHUB_DEV_FIRMWARE,
+							   FU_TYPE_GENESYS_USBHUB_PD_FIRMWARE,
+							   FU_TYPE_GENESYS_USBHUB_CODESIGN_FIRMWARE,
+							   G_TYPE_INVALID);
+		if (firmware_sub == NULL) {
+			g_prefix_error(error, "fw bytes have dual hub firmware: ");
+			return FALSE;
+		}
+		fu_firmware_add_image(firmware, firmware_sub);
```
Comment (adamgene) [inline plugins/genesys/fu-genesys-usbhub-firmware.c:331]: The offset of `FuFirmware` is gone with `FWUPD_INSTALL_FLAG_NO_SEARCH`. Should add `fu_firmware_set_offset(firmware_sub, offset);` here.

**Review by adamgene — CHANGES_REQUESTED** [review]
How an amazing usage with `fu_firmware_new_from_gtypes`! There has a bit error should fix.

File: plugins/genesys/fu-genesys-usbhub-firmware.c (hunk 1 above)
Comment (hughsie) [reply plugins/genesys/fu-genesys-usbhub-firmware.c:331]: Good catch, fixed!

File: libfwupdplugin/fu-firmware.c [hunk 2]
```diff
@@ -912,7 +912,15 @@ fu_firmware_parse_full(FuFirmware *self,
...
-	fu_firmware_set_bytes(self, fw);
+	if (offset == 0x0) {
+		fu_firmware_set_bytes(self, fw);
+	} else {
+		g_autoptr(GBytes) fw_offset = NULL;
+		fw_offset = fu_bytes_new_offset(fw, offset, g_bytes_get_size(fw) - offset, error);
+		if (fw_offset == NULL)
+			return FALSE;
+		fu_firmware_set_bytes(self, fw_offset);
+	}
```
Comment (adamgene) [inline libfwupdplugin/fu-firmware.c:923]: It has conflict with `fu_firmware_set_offset(firmware_sub, offset);`. By this https://github.com/fwupd/fwupd/blob/d7596d3ccec2b0faf701f1849754e0e6a5700c61/plugins/genesys/fu-genesys-usbhub-device.c#L2028-L2031 Which one is better, replace `g_bytes_get_region` or remove the `fu_firmware_set_offset(firmware_sub, offset);` ?

**Review by adamgene — CHANGES_REQUESTED** [review]

File: libfwupdplugin/fu-firmware.c (hunk 2 above)
Comment (hughsie) [reply libfwupdplugin/fu-firmware.c:923]: ohh, i missed that on review -- we shouldn't use g_bytes_get_region as that was introduced in 2.70 and we only depend on GLib 2.68. The fu_bytes_new_offset() change was a valid fix -- it was an oversight as firmware shouldn't have to keep track of the [global] offset in the bytes buffer. Give me a few mins.

Comment (hughsie) [reply libfwupdplugin/fu-firmware.c:923]: Okay, fixed; can you take a look again please.

**Comment by adamgene** [comment]
The parsed result.
```
<firmware gtype="FuGenesysUsbhubFirmware">
  <flags>has-checksum</flags>
  <id>hub</id>
  <version>89.01</version>
  <version_raw>0x8901</version_raw>
  <alignment>0xa</alignment>
  <size>0xb000</size>
  <data size="0x20000">.U~.-.GLHUB.he..</data>
  <alignment>0xa</alignment>
  <tool_string_version>dynamic-13-byte</tool_string_version>
  <mask_project_code>7455</mask_project_code>
  <mask_project_hardware>B</mask_project_hardware>
  <mask_project_firmware>03</mask_project_firmware>
  <mask_project_ic_type>352510</mask_project_ic_type>
  <running_project_code>7455</running_project_code>
  <running_project_hardware>B</running_project_hardware>
  <running_project_firmware>03</running_project_firmware>
  <running_project_ic_type>352510</running_project_ic_type>
  <firmware gtype="FuGenesysUsbhubPdFirmware">
    <flags>has-checksum</flags>
    <id>pd</id>
    <idx>0x2</idx>
    <version>12.00</version>
    <version_raw>0x1200</version_raw>
    <offset>0xb000</offset>
    <alignment>0xa</alignment>
    <size>0xd000</size>
    <data size="0x15000">.{..6....&quot;&quot;..r..</data>
    <alignment>0xa</alignment>
  </firmware>
  <firmware gtype="FuGenesysUsbhubDevFirmware">
    <flags>has-checksum</flags>
    <id>dev-bridge</id>
    <idx>0x1</idx>
    <version>66.11</version>
    <version_raw>0x6611</version_raw>
    <offset>0x18000</offset>
    <alignment>0xa</alignment>
    <size>0x8000</size>
    <data size="0x8000">.G......&quot;</data>
    <alignment>0xa</alignment>
  </firmware>
</firmware>
```

File: plugins/genesys/fu-genesys-usbhub-device.c [hunk 3]
```diff
@@ -2020,32 +2020,23 @@ fu_genesys_usbhub_device_update_firmware(FuGenesysUsbhubDevice *self,
 					 GError **error)
 {
 	gboolean skip_erase = FALSE;
+	gsize bufsz = 0;
+	const guint8 *buf = NULL;
 	const guint start_addr = fu_firmware_get_addr(firmware);
-	const guint code_size = fu_firmware_get_size(firmware);
```
Comment (adamgene) [inline plugins/genesys/fu-genesys-usbhub-device.c:2024]: As the result, the `code_size` is necessary for the firmware.

**Review by adamgene — CHANGES_REQUESTED** [review]
BTW, this also needs to replace https://github.com/fwupd/fwupd/blob/a55351237ba4fa55c54200e150d1fb57013083d7/plugins/genesys/fu-genesys-usbhub-device.c#L1597-L1600

File: plugins/genesys/fu-genesys-usbhub-device.c (hunk 3 above)
Comment (hughsie) [reply plugins/genesys/fu-genesys-usbhub-device.c:2024]: > BTW, this also needs to replace 

I've added a fixup -- can you email me your test firmware please richard_at_hughsie_dot_com and I'll make sure we include the extra sub firmware types in the fuzzing target -- which will also catch things like this.

Comment (adamgene) [reply plugins/genesys/fu-genesys-usbhub-device.c:2024]: Please try the binaries I sent.

Comment (hughsie) [reply plugins/genesys/fu-genesys-usbhub-device.c:2024]: I've tried a few of the firmwares and they seem to have the same putput as before; i.e. the parsing is identical. Can you point me in the right direction please?

Comment (adamgene) [reply plugins/genesys/fu-genesys-usbhub-device.c:2024]:
```
    <size>0xd000</size>
    <data size="0x15000">.{..6....&quot;&quot;..r..</data>
    ```

    The firmware should only write the `size` of data (0xd000), not `data size`(0x15000).

Comment (hughsie) [reply plugins/genesys/fu-genesys-usbhub-device.c:2024]: aha, right, got it; give me 10 mins.

Comment (hughsie) [reply plugins/genesys/fu-genesys-usbhub-device.c:2024]: @adamgene can you try this branch again please?

**Review by adamgene — APPROVED** [review]

**Comment by hughsie** [comment]
Thanks @adamgene -- a super team we make! :)

**Pull request merged**